egs/wsj/s5/steps/nnet3/train_rnn.py

```diff
@@ -0,0 +1,657 @@
+    parser.add_argument("--trainer.optimization.initial-effective-lrate", type=float, dest='initial_effective_lrate',
+                        default = 0.0003,
+                        help="Learning rate used during the initial iteration")
+    parser.add_argument("--trainer.optimization.final-effective-lrate", type=float, dest='final_effective_lrate',
+                        default = 0.00003,
+                        help="Learning rate used during the final iteration")
+    parser.add_argument("--trainer.optimization.num-jobs-initial", type=int, dest='num_jobs_initial',
+                        default = 1,
+                        help="Number of neural net jobs to run in parallel at the start of training")
+    parser.add_argument("--trainer.optimization.num-jobs-final", type=int, dest='num_jobs_final',
+                        default = 8,
+                        help="Number of neural net jobs to run in parallel at the end of training")
+    parser.add_argument("--trainer.optimization.max-models-combine", type=int, dest='max_models_combine',
+                        default = 20,
+                        help = """ The is the maximum number of models we give to the
+                                   final 'combine' stage, but these models will themselves
+                                   be averages of iteration-number ranges. """)
+    parser.add_argument("--trainer.optimization.momentum", type=float, dest='momentum',
+                        default = 0.5,
+                        help="""Momentum used in update computation.
+                        Note: we implemented it in such a way that
+                        it doesn't increase the effective learning rate.""")
+    parser.add_argument("--trainer.optimization.shrink-value", type=float, dest='shrink_value',
+                        default = 0.99,
+                        help="Scaling factor used for scaling the parameter matrices when the derivative averages are below the shrink-threshold at the non-linearities")
+    parser.add_argument("--trainer.optimization.shrink-threshold", type=float, dest='shrink_threshold',
+                        default = 0.15,
+                        help="If the derivative averages are below this threshold we scale the parameter matrices with the shrink-value. It is less than 0.25 for sigmoid non-linearities.")
+
+    # RNN specific trainer options
+    parser.add_argument("--trainer.rnn.num-chunk-per-minibatch", type=int, dest='num_chunk_per_minibatch',
+                        default=100,
+                        help="Number of sequences to be processed in parallel every minibatch" )
+    parser.add_argument("--trainer.rnn.num-bptt-steps", type=int, dest='num_bptt_steps',
+                        default=None,
+                        help="The number of time steps to back-propagate from the last label in the chunk. By default it is same as the chunk-width." )
+    parser.add_argument("--trainer.rnn.shrink", type=float, dest='shrink',
```

Comment:
just a note on shrinking, although it won't affect this pull request for now.
I am working in ~dpovey/kaldi-chain/ on a more elegant mechanism called 'self-repairing nonlinearities'.  It's an option in the nonlinearities (ReLU, sigmoid, tanh) where once the average value or derivative stats (as appropriate) get out of the 'good' range, it adds a small term to the derivative to try to force it back to a good range.  Hopefully this will work better than the shrinking mechanism- certainly it's more general.